### DISCUSSION TITLE
feat: window-step-resize - Wayland

### DISCUSSION THREAD

**Discussion by johnmmcgee** [ghostty-org/ghostty#4074] [Feature Requests, Ideas] [answered]
Hello,

It would be nice to have window-step-resize support in wayland. 

Thank you.

**Comment by rhodes-b** [comment]
Ghostty attempts to have a native GNOME experience using GTK which supports x11 and wayland. I see this as something unlikely to get in as its protocol specific.

If you are in the discord you may be interested to hear that someone is working on a wayland apprt port right now that you might have interest in

**Reply by johnmmcgee** [reply]
I think having the option available for any platform, not just macOS, would be ideal.

**Comment by mitchellh** [comment]
We asked the GTK IRC about this and they said it is not possible with GTK4. The underlying windowing system may allow it but GTK doesn't expose any way for us to take advantage of it.

**Reply by johnmmcgee** [reply]
I wonder if it would be worth submitting as a feature request to gtk team, or if they have a specific reason to not want the ability?

**Comment by pluiedev** [comment, marked as answer]
For Wayland specifically it's not possible for any app to communicate to the compositor that it wants to be step-resized. It's up to the compositor to interpret resize events and make the window step-resized if it wants to, and I'm not aware of any compositors that actually do this. Everything related to window placement and sizes is to be determined by the compositor in Wayland.

**Reply by johnmmcgee** [reply]
Hello,

Is there any reason that you could think of as to why this isn't an issue with gnome-terminal?

**Reply by pluiedev** [reply]
gnome-terminal uses GTK 3 which uses a *lot* of hacks and uses Wayland in ways that it shouldn't be used. Feel free to peruse [`gdkwindow-wayland.c`](https://gitlab.gnome.org/GNOME/gtk/-/blob/gtk-3-24/gdk/wayland/gdkwindow-wayland.c?ref_type=heads) yourself. The core issue is that GTK 3's platform abstraction is still based on X11, which allows things that Wayland was designed to prohibit: e.g. arbitrary window placement and window self-resizing.

As another data point, none of the other GTK 4 terminals (Ptyxis, GNOME Console, etc.) support step resizing, for this exact reason. We're not interested in implementing the hacks the GTK 3 authors had to figure out.
